service/controller/v24/resource/ipam/create.go

```diff
@@ -212,6 +219,63 @@ func (r *Resource) selectRandomAZs(n int) ([]string, error) {
 	return shuffledAZs, nil
 }
 
+// getAvailabilityZonesForInstanceTypes gets availability zones for current
+// region and removes ones that don't have required instance types present.
+func getAvailabilityZonesForInstanceTypes(ctx context.Context, instanceTypes ...string) ([]string, error) {
+	ctlCtx, err := controllercontext.FromContext(ctx)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	var azs []string
+	{
+		// List all available AZs.
+		i := &ec2.DescribeAvailabilityZonesInput{}
+		o, err := ctlCtx.AWSClient.EC2.DescribeAvailabilityZones(i)
+		if err != nil {
+			return nil, microerror.Mask(err)
+		}
+
+		for _, az := range o.AvailabilityZones {
+			azs = append(azs, *az.ZoneName)
+		}
+	}
+
+	// Verify that AZs provide specified instance types.
+	if len(instanceTypes) > 0 {
+		// The mechanism here to find out whether given AZ provides required
+		// instance types or not is really a hack, but as of writing this, AWS
+		// SDK doesn't provide direct way to query available instance types and
+		// indirect use of spot price API for this seems to be the common thing
+		// to do.
+		//
+		// So what happens here is that for each AZ there is a request to getch
+		// EC2 spot instance price history for last minute to see whether the
+		// instance type is available or not.
+		minuteAgo := time.Now().Add(-11 * time.Minute)
+
+		var awsInstanceTypes []*string
+		for _, it := range instanceTypes {
+			awsInstanceTypes = append(awsInstanceTypes, aws.String(it))
+		}
+
+		for _, az := range azs {
+			i := &ec2.DescribeSpotPriceHistoryInput{
+				AvailabilityZone: aws.String(az),
+				InstanceTypes:    awsInstanceTypes,
+				StartTime:        &minuteAgo,
+			}
+
+			_, err := ctlCtx.AWSClient.EC2.DescribeSpotPriceHistory(i)
```

Comment:
Something is missing here, how do we remove the az if there is not Price History?

Reply:
I marked this PR as Draft because I need to figure out what kind of error does the API throw at me when instance is not available. The API documentation is so generous that it doesn't have any mentioning about reasonable errors that can be thrown from there. I'll add that functionality when I know the error better :slightly_smiling_face:

Reply:
Ah okay, makes sense :+1: